lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,468 @@
+/// `Source` node to nodes that are predecessor of it on the spanning tree.
+template<class GraphT, class GT>
+CycleEquivalenceAnalysis<GraphT, GT>::EdgesVectorType
+CycleEquivalenceAnalysis<GraphT, GT>::getTreeEdges(BlockNode *Source) {
+  llvm::SmallVector<EdgeDescriptor> TreeEdges;
+  for (auto &[Neighbor, Index] :
+       llvm::children_edges<llvm::Undirected<BlockNode *>>(Source)) {
+    EdgeDescriptor Edge = EdgeDescriptor(Neighbor, Source);
+    if (SpanningTreeEdges.contains(Index) and not isBackedge(Edge)) {
+      TreeEdges.push_back(Edge);
+    }
+  }
+  return TreeEdges;
+}
+
+// Helper function that creates the `GenericGraph` representing the undirected
+// version of the original CFG
+template<class GraphT, class GT = llvm::GraphTraits<GraphT>>
+static CycleEquivalenceAnalysis<GraphT>::BlockGraph
+initializeGenericGraph(GraphT F) {
+  using NodeT = GT::NodeRef;
+
+  using BlockNode = CycleEquivalenceAnalysis<GraphT>::BlockNode;
+  using BlockGraph = CycleEquivalenceAnalysis<GraphT>::BlockGraph;
+
+  BlockGraph Graph;
+
+  std::map<NodeT, BlockNode *> BBToNode;
+
+  // For every BasicBlock in F, we create a corresponding node in the
+  // `GenericGraph`, and we keep a correspondence between the `BasicBlock` and
+  // the node in the `GenericGraph`
+  auto NodesRange = llvm::make_range(GT::nodes_begin(F), GT::nodes_end(F));
+  for (NodeT BB : NodesRange) {
+    BBToNode[BB] = Graph.addNode(BB);
+  }
+
+  // Set the entry node
+  BlockNode *EntryNode = BBToNode.at(GT::getEntryNode(F));
+  Graph.setEntryNode(EntryNode);
+
+  // We should ensure that we find a single exit node for the `GenericGraph`, so
+  // that we can insert the edge connecting the single exit to the entry node
+  bool ExitConnected = false;
+
+  // Incremental `Label` pool, we reserve the `0` index for the exit->entry edge
+  size_t LabelEdgePool = 1;
+
+  // For every BasicBlock in F, we connect its successors in the `GenericGraph`
+  for (NodeT BB : NodesRange) {
+    BlockNode *From = BBToNode.at(BB);
+
+    // Iterate over the successors
+    auto SuccessorsRange = llvm::make_range(GT::child_begin(BB),
+                                            GT::child_end(BB));
+    for (NodeT SuccessorBB : SuccessorsRange) {
+
+      // Insert the edge from `BB` to `SuccessorBB` and at each edge insertion,
+      // we add an incremental distinct label
+      BlockNode *SuccessorNode = BBToNode.at(SuccessorBB);
+      From->addSuccessor(SuccessorNode, LabelEdgePool);
+      LabelEdgePool++;
+    }
+
+    // TODO: the `exit` -> `entry` backedge insertion should be performed in the
+    //       agnostic part
+    // If the BasicBlock has no successors, we insert an additional and
+    // artificial backedge to the entry node
+    if (SuccessorsRange.empty()) {
+
+      // We should check that a single exit node has been identified at this
+      // point
+      revng_assert(not ExitConnected);
+
+      From->addSuccessor(EntryNode, 0);
+      ExitConnected = true;
+    }
```

Comment:
Leave the the handling of exit -> entry edge at the very end, and out of this loop.

Drop `ExitConnected` and adopt a `SmallSet<NodeT>`, where you put all the nodes without successors.

At the very end, after the loop, you just check that the size of the `SmallSet`is exactly 1. If it's more than 1 the graph has many exits, if it's 0 (that was currently unchecked) the graph doesn't have an exit, which is also wrong. In both cases you can `revng_abort` with a meaningful message, so that whoever is using the API can adjust their graph before computing the cycle equivalence classes.

Reply:
I dropped the `ExitConnected` flag, and went with the solution with the `SmallSet` as you suggested here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L358.

Reply:
With the latest introduced changes, and the possibility of sinking graphs with multiple exit, an explicit handling depending on the number of identified exit nodes has been here https://github.com/revng/revng-c/blob/32e8b34c86eae349bd0b23530e69160e8d8b8b43/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L430